### PR TITLE
Fix onboard sd printing support for onboard filament sensors

### PR DESCRIPTION
**Pull request by Subcircuit**
### Fix onboard sd printing support for onboard filament sensors and add more TFT filament sensors ports.

1. TFT selects onboard sd printing, which can be processed on TFT when Filament runout.
2. Use TFT Unused ports add more  filament sensors support.(PS_ON, USART3, USART4)

### CODE REVIEW

**Comment by digant73** [comment]
does it allow to handle M600 through Marlin FW? if so, what features have to be enabled in Marlin FW (e.g. host commands?)?

**Comment by jedisintraining3** [comment]
with this current code how do you wire up the filament sensor's positive +5 volt port? is it connected to the unused RX or TX port or do you have to connect it directly to one of the board's 5 volt sources?

**Comment by jedisintraining3** [comment]
> with this current code how do you wire up the filament sensor's positive +5 volt port? is it connected to the unused RX or TX port or do you have to connect it directly to one of the board's 5 volt sources?

I figured it out! You can connect the Five volt plug to the Power port on the TFT!

**Comment by oldman4U** [comment]
@jedisintraining3 

To make a certain function working properly is the first step. To make the same function available also for others by documenting it is the next step. Would you be so kind to extend the existing manual at the front page. Thank you

**Comment by Subcircuit** [comment]
@digant73 requires HOST_ACTION_COMMANDS and EMERGENCY_PARSER

**Comment by jedisintraining3** [comment]
> @jedisintraining3
> 
> To make a certain function working properly is the first step. To make the same function available also for others by documenting it is the next step. Would you be so kind to extend the existing manual at the front page. Thank you

How do I edit the manual, If you list the steps, I will greatly add many changes that I think will help many, things that I would have wanted when I was working on this stuff.

**Comment by jedisintraining3** [comment]
I am working to commit a new version of the manual.

**Comment by jedisintraining3** [comment]
I was working on the code and muanual, but I noticed that equipping more than one filament runout sensor in the firmware is a little difficult. You can't just edit the config.ini you have to completely build a new firmware. This would be hard for many non-tech savvy users, and I think that you should add the function of changing the pins to config.ini. Config.ini should have the function of changing pins for ps_on, filament runout(multiple), and btt mini ups.

**Comment by ETE-Design** [comment]
@Subcircuit is HOST_PROMPT_SUPPORT also needed?

**Comment by Subcircuit** [comment]
@ETE-Design  Yse, require uncomment HOST_PROMPT_SUPPORT.  
Through my tests from onboard SD print will not automatic exit after print done, although it has no influence.

**Comment by blueeagle69** [comment]
> @ETE-Design Yse, require uncomment HOST_PROMPT_SUPPORT.
> Through my tests from onboard SD print will not automatic exit after print done, although it has no influence.

So the SD print in TFT mode not ending correctly is still present?

**Comment by bigtreetech** [comment]
> > @ETE-Design Yse, require uncomment HOST_PROMPT_SUPPORT.
> > Through my tests from onboard SD print will not automatic exit after print done, although it has no influence.
> 
> So the SD print in TFT mode not ending correctly is still present?

@blueeagle69  Hello, This problem needs fix by Marlin, which can be fixed here https://github.com/MarlinFirmware/Marlin/pull/19102

**Comment by jedisintraining3** [comment]
Will two encoder's work with this firmware or just two two-state sensors.

**Comment by digant73** [comment]
Will this PR allow the M600 support also printing from TFT's SD card/USB?
Currently in interfaceCmd.c, M600 is intercepted and replaced with a pause. Will this handling be modified in order M600 is routed to main board so Marlin will then start the communication according to this PR?

**Comment by bigtreetech** [comment]
> Will this PR allow the M600 support also printing from TFT's SD card/USB?
> Currently in interfaceCmd.c, M600 is intercepted and replaced with a pause. Will this handling be modified in order M600 is routed to main board so Marlin will then start the communication according to this PR?

M600 will not be replaced to pause if you comment out `#define NOZZLE_PAUSE_M600_M601` in configuration.h

**Comment by digant73** [comment]
> > Will this PR allow the M600 support also printing from TFT's SD card/USB?
> > Currently in interfaceCmd.c, M600 is intercepted and replaced with a pause. Will this handling be modified in order M600 is routed to main board so Marlin will then start the communication according to this PR?
> 
> M600 will not be replaced to pause if you comment out `#define NOZZLE_PAUSE_M600_M601` in configuration.h

Ok, clear.
So, just to recap the features needed on last Marlin FW 2.0.5.6:

Configuratuon.h:
NOZZLE_PARK_FEATURE (it is required by ADVANCED_PAUSE_FEATURE)

Configuration_adv.h:
EMERGENCY_PARSER
ADVANCED_PAUSE_FEATURE
HOST_ACTION_COMMANDS
HOST_PROMPT_SUPPORT

Is there something more to enable?

**Comment by sam01508** [comment]
Not having used two colours yet with my touchscreen in touchscreen mode, is m600 not yet supported? Or will I need Marlin mode?

**Comment by ETE-Design** [comment]
@Subcircuit Seems like Runout placerd on mainboard instead of TFT isen't working :-( At least nothing happen on the screen when it runout... Do we ned to change anything in the TFT Firmware to make it Work? Also, is the M600 working like it should now?

**Comment by blueeagle69** [comment]
Noticed the same thing.
I moved my Smart Filament sensor to the mainboard, all the needed settings are enabled in Marlin..
Yet when my sensor triggers the runout, the printer pauses as expected, but I get no feedback or options on the TFT.
Will have to use Marlin mode I think until this is fixed.

**Comment by paulanashat22** [comment]
i make a simple solution for this issue i khnow that willnot be the perfect one but it's working first 
go to configuration in marlin and make z of nozzle park point 0
  #define NOZZLE_PARK_POINT { (X_MIN_POS + 10), (Y_MIN_POS + 10), 0 } 
when the filament sensor detect that filament runout or(M600 ) the nozzle will go to park point the screen will not give you any data in the touch mode just wait for the extruder until it push the filament up then press the rest button on the screen after that  make preheat and put your new filament and go to print from sd card the tft will ask you to continue printing press okay and it will continue with your new color

**Comment by blueeagle69** [comment]
I take it all back!
My BTT Smart Filament sensor is connected to my SKR1.4 (Biqu B1).
I have the relevant settings in Marlin;
- #define HOST_ACTION_COMMANDS
- #define HOST_PROMPT_SUPPORT 
- #define M114_DETAIL
- #define EMERGENCY_PARSER

I had to start a dummy print in TFT mode to take a photo of my screen for a bug report.
Because I was doing a dummy print, I had forgot to put filament in the sensor.
Next think I know the screen was alerting me.
It showed "Filament Runout!"
I left it a while and then I got "Nozzle Parked".
I clicked to reheat, it fed in filament, asked me if I wanted to purge more, or continue.
Basically everything Marlin  would have been asking and doing.
To say I'm happy is an understatement!

**Comment by ETE-Design** [comment]
@blueeagle69 Great, could you pls. Share the 2 configuration files from Marlin then I can compare with my own thanks :-)

**Comment by blueeagle69** [comment]
No probs;
[Marlin.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/5341208/Marlin.zip)

**Comment by sadbass** [comment]
a little while ago I solved the problem of the layer when I print from the onboard SD now also the filament sensor does not work well, if I knew it had all these problems I did not buy the screen.

**Comment by sadbass** [comment]
@blueeagle69Do you have the TFT connected via rs232 or in emulation mode?

**Comment by blueeagle69** [comment]
Pretty sure it'll be RS232 or whatever way Biqu has it set up in the B1 printer.

But i have managed to have the runout sensor working in Marlin and tft mode when the sensor (Bigtreetech Smart Sensor) is connected to the main board and I print from the onboard SD.
Which I do on my B1 anyway.
